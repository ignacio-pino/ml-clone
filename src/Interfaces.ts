export interface IProductListData {
  id: string;
  title: string;
  price: string;
  original_price?: string;
  currency_id: "UYU" | "USD";
  thumbnail: string;
  shipping: {
    free_shipping: boolean;
  };
}

export interface ISortDataList {
  sort: ISortData;
  available_sorts: ISortData[];
}

export interface ISortData {
  id: string;
  name: string;
}
export interface IProductCard {
  id: string;
  title: string;
  price: string;
  original_price?: string;
  currency_id: "UYU" | "USD";
  thumbnail: string;

  free_shipping: boolean;
}

export interface IFilterDataList {
  filters: IFilterList[];
  availableFilters: IFilterList[];
}
export interface IFilterList {
  id: string;
  name: string;
  type: string;
  values: IFilterListValues[];
}

export interface IFilterListValues {
  id: string;
  name: string;
  results?: number;
  path_from_root?: IPathFromRoot[];
}

export interface ICategoryData {
  query?: string;
  categoryInformation?: ICategoryInformation;
  totalResults: number;
  childrenCategories?: IPathFromRoot[];
}

export interface ICategoryInformation {
  id: string;
  name: string;
  path_from_root?: IPathFromRoot[];
}

export interface IPathFromRoot {
  id: string;
  name: string;
}

export interface IChildrenCategoryData {
  id: string;
  name: string;
  total_items_in_this_category: number;
}

export interface IHeaderCategoryData {
  categoryId: string;
  categoryName: string;
}
