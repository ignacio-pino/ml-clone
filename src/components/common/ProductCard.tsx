import {
  Card,
  Stack,
  CardBody,
  Image,
  Text,
  HStack,
  Tooltip,
  Button,
} from "@chakra-ui/react";
import { IProductCard } from "../../Interfaces";
import { createSearchParams, useNavigate } from "react-router-dom";

export default function ProductCard({
  id,
  title,
  price,
  currency_id,
  original_price,
  free_shipping,
  thumbnail,
}: IProductCard) {
  const navigate = useNavigate();

  const params = { item: id };

  const handleClick = () => {
    navigate({
      pathname: "/item",
      search: `${createSearchParams(params)}`,
    });
  };

  return (
    <Card
      data-testid="productCard"
      id={id}
      direction={{ base: "column", sm: "row" }}
      variant="outline"
      p={1}
      boxShadow={"dark-lg"}
      maxW={"50vw"}
    >
      <Image objectFit="fill" h="130px" w="130px" src={thumbnail} alt={title} />

      <Stack overflow={"hidden"}>
        <CardBody p={3}>
          <Button
            variant={"link"}
            size="lg"
            mt={-2}
            noOfLines={1}
            textOverflow={"ellipsis"}
            onClick={() => handleClick()}
          >
            <Tooltip
              openDelay={500}
              label={title}
              aria-label={title}
              placement="top-end"
            >
              {title}
            </Tooltip>
          </Button>

          {original_price && (
            <HStack gap={0.5} as={"del"} mt={1} ml={0.5}>
              <Text fontSize="x-small" textColor={"gray.500"}>
                {currency_id === "USD" ? "U$S" : "$"}{" "}
                {original_price.split(".")[0]}
              </Text>
              <Text fontSize={"xx-small"} mt={-1} textColor={"gray.500"}>
                {original_price.split(".")[1]}
              </Text>
            </HStack>
          )}
          <HStack mb="2" mt={original_price ? "-2" : "5"}>
            <HStack gap={0.5}>
              <Text as="b" fontSize={"larger"}>
                {currency_id === "USD" ? "U$S" : "$"} {price.split(".")[0]}
              </Text>
              <Text as="b" fontSize={"x-small"} mt={-1}>
                {price.split(".")[1]}
              </Text>
            </HStack>

            {original_price && (
              <Text fontSize={"xs"} textColor={"purple.400"}>
                {Math.floor(
                  ((parseFloat(original_price) - parseFloat(price)) /
                    parseFloat(original_price)) *
                    100
                )}
                % OFF
              </Text>
            )}
          </HStack>
          {free_shipping && (
            <Text
              as="mark"
              fontWeight={"bold"}
              px={2}
              fontSize={"sm"}
              backgroundColor={"purple.400"}
              borderRadius={"4px"}
              textColor={"purple.900"}
            >
              Env√≠o gratis
            </Text>
          )}
        </CardBody>
      </Stack>
    </Card>
  );
}
