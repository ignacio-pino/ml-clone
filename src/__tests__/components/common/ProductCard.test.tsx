import { render, screen } from "@testing-library/react";
import ProductCard from "../../../components/common/ProductCard";
import { IProductCard } from "../../../Interfaces";

describe("ProductCard", () => {
  const sampleProduct: IProductCard = {
    title: "Sample Product",
    price: "100.00",
    currency_id: "UYU",
    original_price: "120.00",
    free_shipping: true,
    thumbnail: "sample-thumbnail-url",
  };

  it("should match snapshot", () => {
    const productCard = render(<ProductCard {...sampleProduct} />).container;

    expect(productCard).toMatchSnapshot();
  });

  it("renders product details correctly", () => {
    render(<ProductCard {...sampleProduct} />);

    const titleElement = screen.getByText("Sample Product");
    const priceElement = screen.getByText("$ 100");
    const originalPriceElement = screen.getByText("$ 120");
    const discountElement = screen.getByText("16% OFF");
    const freeShippingElement = screen.getByText("Envío gratis");
    const thumbnailElement = screen.getByAltText("Sample Product");

    expect(titleElement).toBeInTheDocument();
    expect(priceElement).toBeInTheDocument();
    expect(originalPriceElement).toBeInTheDocument();
    expect(discountElement).toBeInTheDocument();
    expect(freeShippingElement).toBeInTheDocument();
    expect(thumbnailElement).toBeInTheDocument();
  });

  it("renders discount element only when original_price is provided", () => {
    render(<ProductCard {...{ ...sampleProduct, original_price: "" }} />);

    const discountElement = screen.queryByText("16% OFF");
    const originalPriceElement = screen.queryByText("$ 120");

    expect(discountElement).not.toBeInTheDocument();
    expect(originalPriceElement).not.toBeInTheDocument();
  });

  it("renders free shipping element only when free_shipping is true", () => {
    render(<ProductCard {...{ ...sampleProduct, free_shipping: false }} />);

    const freeShippingElement = screen.queryByText("Envío gratis");
    expect(freeShippingElement).not.toBeInTheDocument();
  });

  it("renders currency as U$S when currency_id is USD", () => {
    render(<ProductCard {...{ ...sampleProduct, currency_id: "USD" }} />);

    const priceElement = screen.getByText("U$S 100");
    const originalPriceElement = screen.getByText("U$S 120");

    expect(priceElement).toBeInTheDocument();
    expect(originalPriceElement).toBeInTheDocument();
  });
});
