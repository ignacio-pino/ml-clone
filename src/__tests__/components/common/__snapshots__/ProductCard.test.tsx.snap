// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductCard > should match snapshot 1`] = `
<div>
  <div
    class="chakra-card css-1q6bp6q"
    data-testid="productCard"
    id="Test id"
  >
    <img
      alt="Sample Product"
      class="chakra-image css-1bywttm"
      src="sample-thumbnail-url"
    />
    <div
      class="chakra-stack css-1qstkuz"
    >
      <div
        class="chakra-card__body css-1knigk4"
      >
        <button
          class="chakra-button css-12b0kss"
          type="button"
        >
          <span
            class="css-1baulvz"
            tabindex="0"
          >
            Sample Product
          </span>
        </button>
        <del
          class="chakra-stack css-h9jpb5"
        >
          <p
            class="chakra-text css-1nggxx3"
          >
            $
             
            120
          </p>
          <p
            class="chakra-text css-qc7exc"
          >
            00
          </p>
        </del>
        <div
          class="chakra-stack css-usuplf"
        >
          <div
            class="chakra-stack css-u2zw5"
          >
            <b
              class="chakra-text css-4ww9f7"
            >
              $
               
              100
            </b>
            <b
              class="chakra-text css-to2aqv"
            >
              00
            </b>
          </div>
          <p
            class="chakra-text css-19ff2qr"
          >
            16
            % OFF
          </p>
        </div>
        <mark
          class="chakra-text css-1d599vv"
        >
          Env√≠o gratis
        </mark>
      </div>
    </div>
  </div>
</div>
`;
